docs/RFCS/20171004_sql_sequences.md

```diff
@@ -0,0 +1,421 @@
+  a `oneof`. (and rename `TableDescriptor` to something more generic like
+  `DatabaseObjectDescriptor`)
+  - Advantages: Seems very DRY and unambiguous: the proto would describe a named
+    database object that could either be a table, view, or sequence (or more
+    things in the future).
+  - Disadvantages:
+    - Requires migration of table descriptor protos.
+    - For functions which should only ever be operating on table descriptors,
+      how can we provide a type system guarantee that the
+      `DatabaseObjectDescriptor` they're receiving is a table, not a view or
+      sequence?
+- Optional field on `TableDescriptor` proto (chosen)
+  - Advantages:
+    - Satisfies all requirements with minimal additional code
+  - Disadvantages:
+    - Clutters `TableDescriptor`
+    - Requires checks to be added so that sequences cannot be modified via
+      DML. Checks like these already exist for VIEWs, and shouldn't be hard
+      to add on to.
+
+### Sequence values
+
+- One range per sequence
+  - Advantages:
+    - Spreads load out; no performance bottleneck or single point of failure
+      for all sequences
+    - Already gets created when a `TableDescriptor` is created
+  - Disadvantages:
+    - A lot of overhead to store a single integer
+- Alongside tables
+  - Advantages: Would keep the KV operations for a sequence-using `INSERT`
+    local to one range (until the table splits)
+  - Sequences can be used by multiple tables; in this situation we wouldn't know
+    where to put it.
+- All sequences in one range (chosen)
+  - Advantages:
+    - Avoids overhead of one range per table.
+    - Allows a sequence to be used from multiple tables.
+  - Disadvantages:
+    - Without load-based or manual splitting, this range would be a single point
+      of failure and/or a performance bottleneck. However, with one range per
+      sequence, many applications will fail if the ranges used in a transaction
+      go down anyway.
+- A `system.sequences` table
+  - Advantages:
+    - All those of the "all sequences in one range" approach
+    - Allows easy introspection
+  - Disadvantages
+    - To use `Inc` KV operation (better than `Get` and then `Set` because it
+      requires only one round trip), we'd have to teach `Inc` about SQL's
+      integer encoding, which seems like a separation-of-concerns violation.
+      The introspectability could be achieved by creating a virtual table.
+      (The `information_schema.sequences` table, which we'll be implementing,
+      gets you some of this: all the metadata, just not the value.)
+
+### Impact of not doing this
+
+Users who prefer sequences or who have applications which use them will have
+an obstacle in moving to Cockroach.
+
+## Unresolved questions
+
+### Design / scope issues
+
+- Should we change our `SERIAL` type to be backed by sequences? I don't think
+  we can do this until we make it faster by implementing `CACHE` (see below).
```

Comment:
I think the bigger problem is making a backward incompatible change. SERIAL is of type int32 while our current SERIAL is int64